File: aether/benchmarks/cross-language/java/Skynet.java
// Java Skynet Benchmark
// Based on https://github.com/atemerev/skynet
// Uses ForkJoinPool for recursive parallel decomposition.
// Below the sequential threshold, subtree sums are computed without forking.

import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveTask;

public class Skynet {
    private static final int SEQ_THRESHOLD = 100;

    static class SkynetTask extends RecursiveTask<Long> {
        private final long offset;
        private final long size;

        SkynetTask(long offset, long size) {
            this.offset = offset;
            this.size = size;
        }

        @Override
        protected Long compute() {
            if (size <= SEQ_THRESHOLD) {
                long sum = 0;
                for (long i = 0; i < size; i++) {
                    sum += offset + i;
                }
                return sum;
            }

            long childSize = size / 10;
            long remainder = size - childSize * 10;
            SkynetTask[] children = new SkynetTask[10];
            long childOffset = offset;

            for (int i = 0; i < 10; i++) {
                long cs = childSize + (i < remainder ? 1 : 0);
                children[i] = new SkynetTask(childOffset, cs);
                childOffset += cs;
            }

            for (int i = 1; i < 10; i++) {
                children[i].fork();
            }

            long sum = children[0].compute();
            for (int i = 1; i < 10; i++) {
                sum += children[i].join();
            }
            return sum;
        }
    }

    public static void main(String[] args) {
        String env = System.getenv("BENCHMARK_MESSAGES");
        long numLeaves = env != null ? Long.parseLong(env) : 1000000;

        // Total tree nodes (same formula as all languages for fair comparison)
        long totalNodes = 0;
        for (long n = numLeaves; n >= 1; n /= 10) totalNodes += n;

        ForkJoinPool pool = new ForkJoinPool();

        long start = System.nanoTime();
        long result = pool.invoke(new SkynetTask(0, numLeaves));
        long elapsed = System.nanoTime() - start;

        System.out.println("Sum: " + result);

        if (elapsed > 0) {
            long nsPerMsg = elapsed / totalNodes;
            double throughput = (double) totalNodes / elapsed * 1_000_000_000.0;
            System.out.println("ns/msg:         " + nsPerMsg);
            System.out.printf("Throughput:     %.2f M msg/sec%n", throughput / 1_000_000.0);
        }

        pool.shutdown();
    }
}

File: aether/benchmarks/cross-language/scala/skynet.scala
package bench.skynet

// Scala Akka Skynet Benchmark
// Based on https://github.com/atemerev/skynet
// Recursive actor tree using Akka actors.
// Below the sequential threshold, subtree sums are computed without spawning.

import akka.actor.{Actor, ActorRef, ActorSystem, Props}
import scala.concurrent.{Await, Promise}
import scala.concurrent.duration._

case class Compute(offset: Long, size: Long)
case class SkynetResult(value: Long)

class SkynetNode(parent: ActorRef) extends Actor {
  var pending = 0
  var total: Long = 0L
  val SeqThreshold = 100L

  def receive = {
    case Compute(offset, size) =>
      if (size <= SeqThreshold) {
        var sum = 0L
        var i = 0L
        while (i < size) {
          sum += offset + i
          i += 1
        }
        parent ! SkynetResult(sum)
        context.stop(self)
      } else {
        val childSize = size / 10
        val remainder = size - childSize * 10
        pending = 10
        var childOffset = offset
        var idx = 0
        while (idx < 10) {
          val cs = childSize + (if (idx < remainder) 1 else 0)
          val child = context.actorOf(Props(new SkynetNode(self)))
          child ! Compute(childOffset, cs)
          childOffset += cs
          idx += 1
        }
      }

    case SkynetResult(value) =>
      total += value
      pending -= 1
      if (pending == 0) {
        parent ! SkynetResult(total)
        context.stop(self)
      }
  }
}

class SkynetRoot(numLeaves: Long, promise: Promise[Long]) extends Actor {
  def receive = {
    case SkynetResult(value) =>
      promise.success(value)
      context.stop(self)
  }

  override def preStart(): Unit = {
    val root = context.actorOf(Props(new SkynetNode(self)))
    root ! Compute(0, numLeaves)
  }
}

object SkynetBenchmark extends App {
  val envVal = sys.env.getOrElse("BENCHMARK_MESSAGES", "1000000")
  val numLeaves = envVal.toLong

  // Total tree nodes (same formula as all languages for fair comparison)
  var totalNodes = 0L
  var nn = numLeaves
  while (nn >= 1) { totalNodes += nn; nn /= 10 }

  val system = ActorSystem("skynet")
  val promise = Promise[Long]()

  val start = System.nanoTime()
  system.actorOf(Props(new SkynetRoot(numLeaves, promise)))
  val result = Await.result(promise.future, 60.seconds)
  val elapsed = System.nanoTime() - start

  println(s"Sum: $result")

  if (elapsed > 0) {
    val nsPerMsg = elapsed / totalNodes
    val throughput = totalNodes.toDouble / elapsed * 1e9
    println(s"ns/msg:         $nsPerMsg")
    println(f"Throughput:     ${throughput / 1e6}%.2f M msg/sec")
  }

  system.terminate()
  Await.result(system.whenTerminated, 10.seconds)
}
